### PR TITLE
Proposed 0.4.0 (#92)

### PR DESCRIPTION
**Pull request by chgray**
* CI/CD work, using examples (with diffable a vs. b expected outputs)
* Misc bug fixes
* Improvements to manifested ETW output; in prep for an auto converter
* Deleted 'systemtap' support; it was unused by all and aging
* Moving from binary drops to code only drops (you install dotnet, then build clog; they're better situated to give good guidance for different distros)
* Move to DOTNET 5.0 (6 coming soon)
* Many improvements to variable encoding (install plumbing)
* Many improvements to error handling; making CLOG easier to use
* Improvements for syslog->clog converter 

Co-authored-by: Chris Gray <chgray@microsoft.com>
Co-authored-by: Nick Banks <nibanks@microsoft.com>

### CODE REVIEW

File: .github/workflows/check-clog-linux.yml [hunk 1]
```diff
@@ -0,0 +1,31 @@
+name: CLOG
```
Comment (nibanks) [inline .github/workflows/check-clog-linux.yml:1]: I recommend having unique names for all these.

File: .github/workflows/check-clog-linux.yml [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    steps:
+    - name: Checkout repository
+      uses: actions/checkout@v2
+
+    - run: |
+         sudo apt-add-repository ppa:lttng/stable-2.12
+         sudo apt-get update
+         sudo apt-get install -y liblttng-ust-dev lttng-tools build-essential
+
+    - name: Build CLOG and run Tests
```
Comment (nibanks) [inline .github/workflows/check-clog-linux.yml:26]: Maybe a good idea to separate build and test so it's easier to know what failed when looking at the failed job name?

File: CMakeLists.txt [hunk 3]
```diff
@@ -47,14 +50,15 @@ function(DOT_NET_LIBRARY_BUILD)
...
     list(REMOVE_AT ARGV 0)
 
-    #message(STATUS "---------------------- GENERATING : ${DEPENDENCY_NAME} -------------------")
-    #message(STATUS "PROJECT : ${PROJECT_PATH}")
-    #message(STATUS "PRODUCES : ${DEST_BINARY}")
-    #message(STATUS "OUTPUTDIR : ${CLOG_ARTIFACT_DIR}/${DOTNET_RUNTIME}")
+    ##message(STATUS "---------------------- GENERATING : ${DEPENDENCY_NAME} -------------------")
+    ##message(STATUS "PROJECT : ${PROJECT_PATH}")
+    ##message(STATUS "PRODUCES : ${DEST_BINARY}")
+    ##message(STATUS "OUTPUTDIR : ${CLOG_ARTIFACT_DIR}/${DOTNET_RUNTIME}")
```
Comment (nibanks) [inline CMakeLists.txt:61]: Necessary?

File: defaults/clog.h [hunk 4]
```diff
@@ -42,6 +42,10 @@
 #define CLOG_STRUCT(numFields) CLOG_STRUCT_HELPER(numFields)
 #endif
 
+#define CLOG_ARRAY_HELPER(length, pointer) pointer, length
+#define CLOG_ARRAY(length, pointer) CLOG_ARRAY_HELPER(length, pointer)
+#define CLOG_STRUCT_HELPER(numFields) numFields
+#define CLOG_STRUCT(numFields) CLOG_STRUCT_HELPER(numFields)
```
Comment (nibanks) [inline defaults/clog.h:48]: Why? This is already defined above, no?

File: src/clog/TraceEmitterModules/CLogSTDOUT.cs [hunk 5]
```diff
@@ -249,10 +122,10 @@ public void TraceLineDiscovered(string sourceFile, CLogOutputInfo outputInfo, CL
                         printf += "%u";
                         break;
                     case CLogEncodingType.Int64:
-                        printf += "%lld";
+                        printf += "%ld";
```
Comment (nibanks) [inline src/clog/TraceEmitterModules/CLogSTDOUT.cs:125]: Why did you change these? `ll` == "long long" == 64-bit

File: src/clog/TraceEmitterModules/CLogSTDOUT.cs [hunk 6]
```diff
@@ -302,16 +174,16 @@ public void TraceLineDiscovered(string sourceFile, CLogOutputInfo outputInfo, CL
                 switch (arg.TypeNode.EncodingType)
                 {
                     case CLogEncodingType.Int32:
-                        cast = "(int)";
+                        //cast = "(int)";
```
Comment (nibanks) [inline src/clog/TraceEmitterModules/CLogSTDOUT.cs:177]: Why are all these commented out?

File: .github/workflows/check-clog-linux.yml (hunk 1 above)
Comment (chgray) [reply .github/workflows/check-clog-linux.yml]: agreed; fixed

File: CMakeLists.txt (hunk 3 above)
Comment (chgray) [reply CMakeLists.txt]: agreed; fixed

File: src/clog/TraceEmitterModules/CLogSTDOUT.cs (hunk 5 above)
Comment (chgray) [reply src/clog/TraceEmitterModules/CLogSTDOUT.cs]: agreed; fixed

File: defaults/clog.h (hunk 4 above)
Comment (chgray) [reply defaults/clog.h]: agreed;  I removed some of these - they're there as part of another branch and in this context they're confusing - we'll address them when merging the other branch later on

File: .github/workflows/check-clog-linux.yml (hunk 2 above)
Comment (chgray) [reply .github/workflows/check-clog-linux.yml:26]: I did this intentionally in hopes having them lumped makes "testing at home" easier; I'm not against flipping back down the road, but could we try this for a while?  I'm wanting to lean a lot more on examples that serve as CI/CD tests and would like to make deving smoother than CI/CD

File: src/clog/TraceEmitterModules/CLogSTDOUT.cs (hunk 6 above)
Comment (chgray) [reply src/clog/TraceEmitterModules/CLogSTDOUT.cs:177]: the goal here is to give every opportunity to the compiler to detect casting issues within the code using CLOG;  the casts puts the compiler at a disadvantage.  The comment should be removed in time, but having it helps clean up code that pass incorrect types - could I keep these comments for a bit and evaluate cleaning the comment later?

File: .github/workflows/check-clog-linux.yml (hunk 2 above)
Comment (nibanks) [reply .github/workflows/check-clog-linux.yml:26]: Sure. We can look into it more later.

**Review by nibanks — APPROVED** [review]
